### PR TITLE
[Customer Center] Add API call to get Customer Center config

### PR DESCRIPTION
**Pull request by vegaro**
Adds `Purchases.shared.loadCustomerCenter()` that calls a new backend endpoint that returns the customer center configuration

This API call doesn't exist yet and it will change. This PR is the ground work so that we don't have to wait for the backend to add this API and we can already pretend the API is there.

### CODE REVIEW

**Comment by vegaro** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/RevenueCat/purchases-ios/3933?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#3933** <a href="https://app.graphite.dev/github/pr/RevenueCat/purchases-ios/3933?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#3919** <a href="https://app.graphite.dev/github/pr/RevenueCat/purchases-ios/3919?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3966** <a href="https://app.graphite.dev/github/pr/RevenueCat/purchases-ios/3966?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @vegaro and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: Sources/CustomerCenter/CustomerCenterConfigData.swift [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+//
+//
+//  Created by Cesar de la Vega on 28/5/24.
+//
+
+import Foundation
+
+// swiftlint:disable missing_docs
+// swiftlint:disable nesting
+public struct CustomerCenterConfigData {
```
Comment (tonidero) [inline Sources/CustomerCenter/CustomerCenterConfigData.swift:20]: We will need to add all these to the API tests

**Review by tonidero — APPROVED** [review]
LGTM! I would probably hold off merging back to main until the feature is ready in case the public API changes (or we can make it internal until everything is ready).

**Review by tonidero — COMMENTED** [review]
Also, we are missing tests? Not sure if we are doing those in a separate PR though

File: Tests/UnitTests/CustomerCenter/CustomerCenterConfigDataTests.swift [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                            .init(
+                                id: "path3",
+                                title: "Path 3",
+                                type: .changePlans,
+                                promotionalOffer: nil,
+                                feedbackSurvey: .init(title: "survey title",
+                                                      options: [
+                                                        .init(id: "id_1",
+                                                              title: "option 1",
+                                                              promotionalOffer: .init(iosOfferId: "offer_id_1"))
```
Comment (vegaro) [inline Tests/UnitTests/CustomerCenter/CustomerCenterConfigDataTests.swift:62]: I added `promotionalOffer` to `Option` in another PR #3968

File: Sources/CustomerCenter/CustomerCenterConfigData.swift [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+            let accentColor: String
+            let backgroundColor: String
+            let textColor: String
+
+            public init(accentColor: String, backgroundColor: String, textColor: String) {
+                self.accentColor = accentColor
+                self.backgroundColor = backgroundColor
+                self.textColor = textColor
+            }
```
Comment (vegaro) [inline Sources/CustomerCenter/CustomerCenterConfigData.swift:151]: There are changes to this struct in #3983 which makes it use `PaywallColor`, so in this PR it's just strings and a mode enum for simplification.

File: Sources/CustomerCenter/CustomerCenterConfigData.swift [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+
+        }
+
+        public enum AppearanceMode: String {
+
+            case custom = "CUSTOM"
+            case system = "SYSTEM"
```
Comment (vegaro) [inline Sources/CustomerCenter/CustomerCenterConfigData.swift:158]: In #3983 there are changes to this enum so this is just a simple String based enum

**Review by tonidero — APPROVED** [review]
Left some comments but nothing really blocking. Only thing might be deciding whether we want to make the public API for `screens` a dictionary, the same as the response.

File: Sources/CustomerCenter/CustomerCenterConfigData.swift [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+//  Created by Cesar de la Vega on 28/5/24.
+//
+
+import Foundation
+
+// swiftlint:disable missing_docs
+// swiftlint:disable nesting
+public struct CustomerCenterConfigData {
+
+    public let screens: [Screen]
```
[the remark this answers is not in the document]
Comment (vegaro) [reply Sources/CustomerCenter/CustomerCenterConfigData.swift:22]: Right, I had previously added a subscript but now that the response is a dict as well, I just made it a dict

File: Sources/Networking/HTTPClient/HTTPRequestPath.swift [hunk 6]
```diff
@@ -135,7 +138,8 @@ extension HTTPRequest.Path: HTTPRequestPath {
                 .postReceiptData,
                 .health,
                 .getOfferings,
-                .getProductEntitlementMapping:
+                .getProductEntitlementMapping,
+                .getCustomerCenterConfig:
```
[the remark this answers is not in the document]
Comment (vegaro) [reply Sources/Networking/HTTPClient/HTTPRequestPath.swift:142]: oh right, good catch, added this long time ago and forgot about it

**Pull request merged**